protocols/relay/src/v2/client.rs

```diff
@@ -0,0 +1,479 @@
+                }
+                RelayedConnection::Operational { substream, .. } => {
+                    return Pin::new(substream).poll_write(cx, buf);
+                }
+            }
+        }
+    }
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Result<(), Error>> {
+        loop {
+            match self.deref_mut() {
+                RelayedConnection::InboundAccepting { accept } => {
+                    *self = ready!(accept.poll_unpin(cx))?;
+                }
+                RelayedConnection::Operational { substream, .. } => {
+                    return Pin::new(substream).poll_flush(cx);
+                }
+            }
+        }
+    }
+    fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Result<(), Error>> {
+        loop {
+            match self.deref_mut() {
+                RelayedConnection::InboundAccepting { accept } => {
+                    *self = ready!(accept.poll_unpin(cx))?;
+                }
+                RelayedConnection::Operational { substream, .. } => {
+                    return Pin::new(substream).poll_close(cx);
+                }
+            }
+        }
+    }
+
+    fn poll_write_vectored(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context,
+        bufs: &[IoSlice],
+    ) -> Poll<Result<usize, Error>> {
+        loop {
+            match self.deref_mut() {
+                RelayedConnection::InboundAccepting { accept } => {
+                    *self = ready!(accept.poll_unpin(cx))?;
+                }
+                RelayedConnection::Operational { substream, .. } => {
+                    return Pin::new(substream).poll_write_vectored(cx, bufs);
+                }
+            }
+        }
+    }
+}
+
+impl AsyncRead for RelayedConnection {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<Result<usize, Error>> {
+        loop {
+            match self.deref_mut() {
+                RelayedConnection::InboundAccepting { accept } => {
+                    *self = ready!(accept.poll_unpin(cx))?;
+                }
+                RelayedConnection::Operational {
+                    read_buffer,
+                    substream,
+                    ..
+                } => {
+                    if !read_buffer.is_empty() {
+                        let n = std::cmp::min(read_buffer.len(), buf.len());
+                        let data = read_buffer.split_to(n);
+                        buf[0..n].copy_from_slice(&data[..]);
+                        return Poll::Ready(Ok(n));
+                    }
+
+                    return Pin::new(substream).poll_read(cx, buf);
+                }
+            }
+        }
+    }
+}
+
+enum RqstPendingConnection {
+    Reservation {
+        to_listener: Sender<transport::ToListenerMsg>,
+    },
+    Circuit {
+        dst_peer_id: PeerId,
+        send_back: oneshot::Sender<Result<RelayedConnection, ()>>,
+    },
+}
```

Comment:
What do you think of this naming instead? Would this describe the same intent or am I missing something?

```suggestion
enum OnNewConnection {
    MakeReservation {
        to_listener: Sender<transport::ToListenerMsg>,
    },
    EstablishCircuit {
        dst_peer_id: PeerId,
        send_back: oneshot::Sender<Result<RelayedConnection, ()>>,
    },
}
```

Reply:
>  Would this describe the same intent or am I missing something?

Yes. I prefer your wording.

Though with https://github.com/libp2p/rust-libp2p/pull/2191 we can actually get rid of the entire command buffering logic.

@thomas what do you think of 26aaa6e28f55a4bb8961bb67dcf3945a0c4f0a72?

Reply:
> @thomas what do you think of [26aaa6e](https://github.com/libp2p/rust-libp2p/commit/26aaa6e28f55a4bb8961bb67dcf3945a0c4f0a72)?

I like it! Left a comment on the commit: https://github.com/libp2p/rust-libp2p/commit/26aaa6e28f55a4bb8961bb67dcf3945a0c4f0a72#r61851854 (not sure if you got a notification for that)

Reply:
For the record: https://github.com/libp2p/rust-libp2p/pull/2059#commitcomment-61954555